### PR TITLE
Document and test _get_packed_offsets()

### PR DESCRIPTION
**Pull request by timhoffm**
## PR Summary

Properly test and document `matplotlib.offsetbox._get_packed_offsets()`.

## PR Checklist

- [x] Has Pytest style unit tests
- [x] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [x] Documentation is sphinx and numpydoc compliant

### CODE REVIEW

File: lib/matplotlib/offsetbox.py [hunk 1]
```diff
@@ -81,6 +117,9 @@ def _get_packed_offsets(wd_list, total, sep, mode="fixed"):
     elif mode == "equal":
         maxh = max(w_list)
         if total is None:
+            if sep is None:
+                raise ValueError("total and sep cannot both be None when "
```
Comment (timhoffm) [inline lib/matplotlib/offsetbox.py]: This raised a TypeError before in the line below (int + None).

Since the function is private, we can do the change without announcement.

File: lib/matplotlib/offsetbox.py [hunk 2]
```diff
@@ -1,6 +1,6 @@
 """
-The OffsetBox is a simple container artist. The child artist are meant
-to be drawn at a relative position to its parent.  The [VH]Packer,
+The `.OffsetBox` is a simple container artist. It's contained children are
```
Comment (anntzer) [inline lib/matplotlib/offsetbox.py]: Its

File: lib/matplotlib/offsetbox.py [hunk 3]
```diff
@@ -49,12 +49,48 @@ def _get_packed_offsets(wd_list, total, sep, mode="fixed"):
     *mode*. xdescent is analogous to the usual descent, but along the
     x-direction. xdescent values are currently ignored.
 
-    *wd_list* : list of (width, xdescent) of boxes to be packed.
-    *sep* : spacing between boxes
-    *total* : Intended total length. None if not used.
-    *mode* : packing mode. 'fixed', 'expand', or 'equal'.
+    For simplicity of the description, the terminology used here insinuates a
```
Comment (anntzer) [inline lib/matplotlib/offsetbox.py]: insinuates -> assumes

File: lib/matplotlib/offsetbox.py [hunk 4]
```diff
@@ -49,12 +49,48 @@ def _get_packed_offsets(wd_list, total, sep, mode="fixed"):
...
+
+    - 'expand': Distribute the boxes with equal spacing so that the left edge
+      of the first box is at 0, and the right edge of the last box is at
+      *total*. The parameter *sep* is ignored in this mode. A total of *None*
+      is accepted and considered equal to 1. The total is returned unchanged
+      (except for the conversion *None* to 1). If the total is smaller than
+      the sum of the widths, the laid out boxed will overlap.
+
+    - 'equal': If *total* is given, the total space is divided in N equal
+      ranges and each box is left-aligned within it's subspace.
```
Comment (anntzer) [inline lib/matplotlib/offsetbox.py]: its

File: lib/matplotlib/offsetbox.py [hunk 5]
```diff
@@ -49,12 +49,48 @@ def _get_packed_offsets(wd_list, total, sep, mode="fixed"):
...
+      right edge of the last box. A non-None total will be passed unchecked
+      to the output. In particular this means that right edge of the last
+      box will may be further to the right than the returned total.
+
+    - 'expand': Distribute the boxes with equal spacing so that the left edge
+      of the first box is at 0, and the right edge of the last box is at
+      *total*. The parameter *sep* is ignored in this mode. A total of *None*
+      is accepted and considered equal to 1. The total is returned unchanged
+      (except for the conversion *None* to 1). If the total is smaller than
+      the sum of the widths, the laid out boxed will overlap.
```
Comment (QuLogic) [inline lib/matplotlib/offsetbox.py]:
```suggestion
      the sum of the widths, the laid out boxes will overlap.
```

File: lib/matplotlib/tests/test_offsetbox.py [hunk 6]
```diff
@@ -123,4 +125,58 @@ def test_get_packed_offsets(wd_list, total, sep, mode):
     # issue tickets (at least #10476 and #10784) related to corner cases
     # triggered inside this function when calling higher-level functions
     # (e.g. `Axes.legend`).
+    # These are just some additional smoke tests. The output is untested.
```
Comment (anntzer) [inline lib/matplotlib/tests/test_offsetbox.py]: It looks like this test could just be deleted as it is essentially covered by the tests below?

Comment (timhoffm) [reply lib/matplotlib/tests/test_offsetbox.py]: Not exactly, e.g. I do not test negative `total` or `sep`. I agree that they'd better be tested explicitly for their output, but until then this is better than nothing.

Since it is not documented what *should* happen in this cases, I've left this as is (unresolved). IMHO that discussion is worth some thought and a separate PR.

File: lib/matplotlib/offsetbox.py [hunk 7]
```diff
@@ -49,12 +49,48 @@ def _get_packed_offsets(wd_list, total, sep, mode="fixed"):
...
-    *total* : Intended total length. None if not used.
-    *mode* : packing mode. 'fixed', 'expand', or 'equal'.
+    For simplicity of the description, the terminology used here assumes a
+    horizontal layout, but the function works equally for a vertical layout.
+
+    There are three packing modes:
+
+    - 'fixed': The elements are packed tight to the left with a spacing of
+      *sep* in between. If *total* is *None* the returned total will be the
+      right edge of the last box. A non-None total will be passed unchecked
```
Comment (QuLogic) [inline lib/matplotlib/offsetbox.py]:
```suggestion
      right edge of the last box. A non-*None* total will be passed unchecked
```

File: lib/matplotlib/offsetbox.py [hunk 8]
```diff
@@ -49,12 +49,48 @@ def _get_packed_offsets(wd_list, total, sep, mode="fixed"):
...
+      to the output. In particular this means that right edge of the last
+      box will may be further to the right than the returned total.
```
Comment (QuLogic) [inline lib/matplotlib/offsetbox.py]:
```suggestion
      box may be further to the right than the returned total.
```

File: lib/matplotlib/offsetbox.py [hunk 9]
```diff
@@ -49,12 +49,48 @@ def _get_packed_offsets(wd_list, total, sep, mode="fixed"):
...
+      ranges and each box is left-aligned within its subspace.
+      Otherwise (*total* is *None*), *sep* must be provided and each box is
+      left-aligned in its subspace of width ``(max(widths) + sep)``. The
+      total width is then calculated to be ``N * (max(widths) + sep)``.
+
+    Parameters
+    ----------
+    wd_list : list of (float, float)
+        (width, xdescent) of boxes to be packed.
+    total : float or None
```
Comment (QuLogic) [inline lib/matplotlib/offsetbox.py]:
```suggestion
    total : float or *None*
```

File: lib/matplotlib/offsetbox.py [hunk 10]
```diff
@@ -49,12 +49,48 @@ def _get_packed_offsets(wd_list, total, sep, mode="fixed"):
...
+        Intended total length. None if not used.
```
Comment (QuLogic) [inline lib/matplotlib/offsetbox.py]:
```suggestion
        Intended total length. *None* if not used.
```

Comment (timhoffm) [reply lib/matplotlib/offsetbox.py]: Generally, I don't think we style parameter types.

File: lib/matplotlib/offsetbox.py (hunk 9 above)
Comment (timhoffm) [reply lib/matplotlib/offsetbox.py]: Generally, I don't think we style parameter types.

File: lib/matplotlib/offsetbox.py (hunk 10 above)
Comment (QuLogic) [reply lib/matplotlib/offsetbox.py]: This one's not a parameter type.
